File: src/GameCore/Unit/EmotionSystem.cs
using GameCore.EventBus;

namespace GameCore.Unit;

/// <summary>情感变更定义（emotionTriggers.json 的条目）。</summary>
public sealed record EmotionChange(string EmotionId, int Intensity);

/// <summary>
/// EmotionSystem — 情感动态管理。
/// 四维数值（恐惧/愤怒/悲伤/喜悦）0-100，> 50 阈值才成为活跃 EMOTION 标签。
/// 跷跷板衰减：提升一种情感会压制对立情感。
/// 订阅属主 = 本实例（不会被 Unit.RecalculateTags 的 unsubscribeAll(this) 清理）。
/// </summary>
public sealed class EmotionSystem
{
    public const string Fear = "恐惧";
    public const string Anger = "愤怒";
    public const string Sorrow = "悲伤";
    public const string Joy = "喜悦";

    private readonly Unit _owner;
    private int _fearLevel;     // 0-100
    private int _angerLevel;    // 0-100
    private int _sorrowLevel;   // 0-100
    private int _joyLevel;      // 0-100
    private readonly HashSet<string> _survivalEmotions = new();   // 生存状态标签（饥饿/干渴/冻伤）

    public int FearLevel => _fearLevel;
    public int AngerLevel => _angerLevel;
    public int SorrowLevel => _sorrowLevel;
    public int JoyLevel => _joyLevel;

    public EmotionSystem(Unit owner, IEventBus bus)
    {
        _owner = owner;
        RegisterListeners(bus);
    }

    private void RegisterListeners(IEventBus bus)
    {
        bus.SubscribeWithOwner(EventTypes.DamageCrit, e =>
        {
            if (e.Target == _owner) ApplyEmotion(Anger, 40);
        }, this);
        bus.SubscribeWithOwner(EventTypes.AllyDeath, _ => ApplyEmotion(Sorrow, 60), this);
        bus.SubscribeWithOwner(EventTypes.Surrounded, e =>
        {
            if (e.Target == _owner) ApplyEmotion(Fear, 30);
        }, this);
        bus.SubscribeWithOwner(EventTypes.EnemyKilled, e =>
        {
            if (ReferenceEquals(e.Source, _owner) || e.Target == _owner) ApplyEmotion(Joy, 20);
        }, this);
        bus.SubscribeWithOwner(EventTypes.HpBelow30, e =>
        {
            if (e.Target == _owner || e.Unit == _owner) ApplyEmotion(Fear, 50);
        }, this);
    }

    /// <summary>提升情感强度并跷跷板衰减对立情感。调用方负责 recalculateTags。</summary>
    public void ApplyEmotion(string emotionId, int intensity)
    {
        switch (emotionId)
        {
            case Fear: _fearLevel = Math.Min(100, _fearLevel + intensity); break;
            case Anger: _angerLevel = Math.Min(100, _angerLevel + intensity); break;
            case Sorrow: _sorrowLevel = Math.Min(100, _sorrowLevel + intensity); break;
            case Joy: _joyLevel = Math.Min(100, _joyLevel + intensity); break;
        }
        DecayOpposing(emotionId, intensity / 2);
    }

    private void DecayOpposing(string emotionId, int amount)
    {
        switch (emotionId)
        {
            case Anger: _joyLevel = Math.Max(0, _joyLevel - amount); break;
            case Joy: _angerLevel = Math.Max(0, _angerLevel - amount); break;
            case Fear: _angerLevel = Math.Max(0, _angerLevel - amount / 2); break;
        }
    }

    /// <summary>每回合自然衰减，返回是否有活跃情感数量变化（true → 需要 RecalculateTags）。</summary>
    public bool TickDecay()
    {
        var before = ActiveEmotionIds().Count;
        _fearLevel = Math.Max(0, _fearLevel - 5);
        _angerLevel = Math.Max(0, _angerLevel - 5);
        _sorrowLevel = Math.Max(0, _sorrowLevel - 8);
        _joyLevel = Math.Max(0, _joyLevel - 5);
        return ActiveEmotionIds().Count != before;
    }

    /// <summary>生存状态标签挂接/移除（饥饿/干渴/冻伤，由生存系统临界值判定驱动）。</summary>
    public void SetSurvivalEmotion(string tagId, bool active)
    {
        if (active) _survivalEmotions.Add(tagId);
        else _survivalEmotions.Remove(tagId);
    }

    /// <summary>> 50 阈值的情感才成为活跃 EMOTION 标签（含生存状态标签）。</summary>
    public HashSet<string> ActiveEmotionIds()
    {
        var ids = new HashSet<string>(_survivalEmotions);
        if (_fearLevel > 50) ids.Add(Fear);
        if (_angerLevel > 50) ids.Add(Anger);
        if (_sorrowLevel > 50) ids.Add(Sorrow);
        if (_joyLevel > 50) ids.Add(Joy);
        return ids;
    }
}

File: canglan-backend/canglan-world/src/main/java/com/canglan/world/unit/EmotionSystem.java
package com.canglan.world.unit;

import com.canglan.core.eventbus.Event;
import com.canglan.core.eventbus.EventBus;
import com.canglan.core.eventbus.EventTypes;

/**
 * EmotionSystem — 情感动态管理。对应 C# EmotionSystem。
 * 四维数值（恐惧/愤怒/悲伤/喜悦）0-100，> 50 阈值才成为活跃 EMOTION 标签。
 * 跷跷板衰减：提升一种情感会压制对立情感。
 * 订阅属主 = 本实例（不会被 Unit.recalculateTags 的 unsubscribeAll(this) 清理）。
 */
public final class EmotionSystem {

    public static final String FEAR = "恐惧";
    public static final String ANGER = "愤怒";
    public static final String SORROW = "悲伤";
    public static final String JOY = "喜悦";

    private final Unit owner;
    private int fearLevel;      // 0-100
    private int angerLevel;     // 0-100
    private int sorrowLevel;    // 0-100
    private int joyLevel;       // 0-100
    private final java.util.HashSet<String> survivalEmotions = new java.util.HashSet<>();  // 生存状态标签（饥饿/干渴/冻伤）

    public int fearLevel() { return fearLevel; }
    public int angerLevel() { return angerLevel; }
    public int sorrowLevel() { return sorrowLevel; }
    public int joyLevel() { return joyLevel; }

    public EmotionSystem(Unit owner, EventBus bus) {
        this.owner = owner;
        registerListeners(bus);
    }

    private void registerListeners(EventBus bus) {
        bus.subscribeWithOwner(EventTypes.DAMAGE_CRIT, e -> {
            if (e.get("target") == owner) applyEmotion(ANGER, 40);
        }, this);
        bus.subscribeWithOwner(EventTypes.ALLY_DEATH, e -> applyEmotion(SORROW, 60), this);
        bus.subscribeWithOwner(EventTypes.SURROUNDED, e -> {
            if (e.get("target") == owner) applyEmotion(FEAR, 30);
        }, this);
        bus.subscribeWithOwner(EventTypes.ENEMY_KILLED, e -> {
            if (e.source() == owner || e.get("target") == owner) applyEmotion(JOY, 20);
        }, this);
        bus.subscribeWithOwner(EventTypes.HP_BELOW_30, e -> {
            Object unit = e.get("unit");
            if (e.get("target") == owner || unit == owner) applyEmotion(FEAR, 50);
        }, this);
    }

    /** 提升情感强度并跷跷板衰减对立情感。调用方负责 recalculateTags。 */
    public void applyEmotion(String emotionId, int intensity) {
        switch (emotionId) {
            case FEAR -> fearLevel = Math.min(100, fearLevel + intensity);
            case ANGER -> angerLevel = Math.min(100, angerLevel + intensity);
            case SORROW -> sorrowLevel = Math.min(100, sorrowLevel + intensity);
            case JOY -> joyLevel = Math.min(100, joyLevel + intensity);
            default -> { }
        }
        decayOpposing(emotionId, intensity / 2);
    }

    private void decayOpposing(String emotionId, int amount) {
        switch (emotionId) {
            case ANGER -> joyLevel = Math.max(0, joyLevel - amount);
            case JOY -> angerLevel = Math.max(0, angerLevel - amount);
            case FEAR -> angerLevel = Math.max(0, angerLevel - amount / 2);
            default -> { }
        }
    }

    /** 每回合自然衰减，返回是否有活跃情感数量变化（true → 需要 recalculateTags）。 */
    public boolean tickDecay() {
        int before = activeEmotionIds().size();
        fearLevel = Math.max(0, fearLevel - 5);
        angerLevel = Math.max(0, angerLevel - 5);
        sorrowLevel = Math.max(0, sorrowLevel - 8);
        joyLevel = Math.max(0, joyLevel - 5);
        return activeEmotionIds().size() != before;
    }

    /** 生存状态标签挂接/移除（饥饿/干渴/冻伤，由生存系统临界值判定驱动）。 */
    public void setSurvivalEmotion(String tagId, boolean active) {
        if (active) survivalEmotions.add(tagId);
        else survivalEmotions.remove(tagId);
    }

    /** > 50 阈值的情感才成为活跃 EMOTION 标签（含生存状态标签）。 */
    public java.util.HashSet<String> activeEmotionIds() {
        java.util.HashSet<String> ids = new java.util.HashSet<>(survivalEmotions);
        if (fearLevel > 50) ids.add(FEAR);
        if (angerLevel > 50) ids.add(ANGER);
        if (sorrowLevel > 50) ids.add(SORROW);
        if (joyLevel > 50) ids.add(JOY);
        return ids;
    }

    /** 读档恢复情感四维。 */
    public void restore(int fear, int anger, int sorrow, int joy) {
        fearLevel = clamp(fear);
        angerLevel = clamp(anger);
        sorrowLevel = clamp(sorrow);
        joyLevel = clamp(joy);
    }

    private static int clamp(int v) { return Math.max(0, Math.min(100, v)); }
}
